lib/Trash/TrashBackend.php

```diff
@@ -313,4 +321,36 @@ public function getTrashNodeById(IUser $user, int $fileId) {
 			return null;
 		}
 	}
+
+	public function expire(Expiration $expiration): array {
+		$size = 0;
+		$count = 0;
+		$folders = $this->folderManager->getAllFolders();
+		foreach ($folders as $folder) {
+			$folderId = $folder['id'];
+			$trashItems = $this->trashManager->listTrashForFolders([$folderId]);
+			$trashFolder = $this->getTrashFolder($folderId);
+			foreach ($trashItems as $groupTrashItem) {
+				if ($expiration->isExpired($groupTrashItem['deleted_time'])) {
+					try {
+						$node = $trashFolder->get($groupTrashItem['name'] . '.d' . $groupTrashItem['deleted_time']);
+						$size += $node->getSize();
+						$count += 1;
+					} catch (NotFoundException $e) {
+						$this->trashManager->removeItem($folderId, $groupTrashItem['name'], $groupTrashItem['deleted_time']);
+						continue;
+					}
+					if ($node->getStorage()->unlink($node->getInternalPath()) === false) {
+						throw new \Exception('Failed to remove item from trashbin');
+					}
+					$node->getStorage()->getCache()->remove($node->getInternalPath());
+					$this->trashManager->removeItem($folderId, $groupTrashItem['name'], $groupTrashItem['deleted_time']);
+					$this->versionsBackend->deleteAllVersionsForFile($folderId, $groupTrashItem['file_id']);
```

Comment:
This is going to cause an exception if file_id is NULL. In my installation this was true for about 25% of files, so it seems to be a relevant case.

```suggestion
					if (!is_null($groupTrashItem['file_id'])) {
						$this->versionsBackend->deleteAllVersionsForFile($folderId, $groupTrashItem['file_id']);
					}

```

Reply:
Ah right, the file_id was added at a later point https://github.com/nextcloud/groupfolders/blob/7d7c1247c59852f6e88f21ba1da07ad1c72a4c41/lib/Migration/Version501000Date20191218182434.php

Good catch.